hadoop-common-project/hadoop-auth/src/main/java/org/apache/hadoop/util/SubjectUtil.java

```diff
@@ -0,0 +1,232 @@
+    }
+  }
+
+  /**
+   * Maps action to a Callable, and delegates to callAs(). On older JVMs, the
+   * action may be double wrapped (into Callable, and then back into
+   * PrivilegedAction).
+   *
+   * @param subject the subject this action runs as
+   * @param action action  the action to run
+   * @return the result of the action
+   * @param <T> the type of the result
+   */
+  public static <T> T doAs(Subject subject, PrivilegedAction<T> action) {
+    try {
+      return callAs(subject, privilegedActionToCallable(action));
+    } catch (CompletionException ce) {
+      Throwable cause = ce.getCause();
+      if (cause != null) {
+        throw sneakyThrow(cause);
+      } else {
+        // This should never happen, as CompletionException should always wrap an exception
+        throw ce;
+      }
+    }
+  }
+
+  /**
+   * Maps action to a Callable, and delegates to callAs(). On older JVMs, the
+   * action may be double wrapped (into Callable, and then back into PrivilegedAction).
+   *
+   * @param subject the subject this action runs as
+   * @param action action  the action to run
+   * @return the result of the action
+   * @param <T> the type of the result
+   * @throws PrivilegedActionException if {@code action.run()} throws an exception.
+   *      The cause of the {@code PrivilegedActionException} is set to the exception
+   *      thrown by {@code action.run()}.
+   */
+  public static <T> T doAs(
+      Subject subject, PrivilegedExceptionAction<T> action) throws PrivilegedActionException {
+    try {
+      return callAs(subject, privilegedExceptionActionToCallable(action));
+    } catch (CompletionException ce) {
+      try {
+        Exception cause = (Exception) ce.getCause();
+        throw new PrivilegedActionException(cause);
+      } catch (ClassCastException castException) {
+        // This should never happen, as PrivilegedExceptionAction should not wrap
+        // non-checked exceptions
+        throw new PrivilegedActionException(new UndeclaredThrowableException(ce.getCause()));
+      }
+    }
+  }
+
+  /**
+   * Maps to Subject.current() if available, otherwise maps to Subject.getSubject().
+   *
+   * @return the current subject
+   */
+  public static Subject current() {
+    try {
+      return (Subject) CURRENT.invoke();
+    } catch (Throwable t) {
+      throw sneakyThrow(t);
+    }
+  }
+
+  @SuppressWarnings("unused")
+  private static <T> PrivilegedExceptionAction<T> callableToPrivilegedExceptionAction(
+      Callable<T> callable) {
+    return callable::call;
+  }
+
+  private static <T> Callable<T> privilegedExceptionActionToCallable(
+      PrivilegedExceptionAction<T> action) {
+    return action::run;
+  }
+
+  private static <T> Callable<T> privilegedActionToCallable(
+      PrivilegedAction<T> action) {
+    return action::run;
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <E extends Throwable> RuntimeException sneakyThrow(Throwable e) throws E {
+    throw (E) e;
```

Comment:
Type checking should be performed here. For `RuntimeException` and `Error`, they should be thrown directly, while other exceptions should be wrapped into a `RuntimeException` and then thrown.

Additionally, does this method need to specify a return value type?

Reply:
the intention here is to allow the caller method to throw the checked exception without explicitly exception list declaration
https://www.baeldung.com/java-sneaky-throws